File: Algotoolbox/week2/fib_huge.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

ll pisano_period(ll m)
{
    int prev = 0, current = 1, temp;
    for (int i = 0; i < (m * m); i++)
    {
        temp = prev;
        prev = current;

        current = (temp + current) % m;

        if (prev == 0 && current == 1)
        {
            return i + 1;
        }
    }
    return -1; //will never reach
}
ll fib_huge(ll n, ll m)
{
    int pisano = (n % pisano_period(m));
    ll prev = 0, cur = 1,temp;
    if (pisano == 0)
        return 0;
    else if (pisano == 1)
        return 1;

    for (int i = 0; i < pisano - 1; ++i)
    {
        temp = prev;
        prev = cur;
        cur = (temp + cur) % m;
    }

    return cur;
}
int main()
{
    ll n, m;
    cin >> n >> m;
    cout << fib_huge(n, m);
    return 0;
}

File: Algotoolbox/week2/fib_huge.py
def pisanoPeriod(m):
    prev, curr = 0,1
    for i in range(0,m*m):
        prev, curr /= curr, (prev+curr)%m;

        if(prev ==0 and curr ==1):
            return i+1

def fib_huge(n,m):
    pisano_per = pisanoPeriod(m)
    n = n % pisano_per
    pre, cu =0,1
    if n==0:
        return 0
    if n==1:
        return 1
    
    for i in range(n-1):
        pre,cu /=cu , pre + cu

    return(cu%m);


if __name__ == "__main__":
    n = 239
    m =1000
    print(fib_huge(n,m))